connectors/services/job_scheduler.py

```diff
@@ -0,0 +1,200 @@
+                    logger.error(e)
+                    return
+                except ServiceTypeNotSupportedError:
+                    logger.debug(
+                        f"Can't handle source of type {connector.service_type}"
+                    )
+                    return
+                except DataSourceError as e:
+                    await connector.error(e)
+                    logger.critical(e, exc_info=True)
+                    self.raise_if_spurious(e)
+                    return
+
+            try:
+                # the heartbeat is always triggered
+                connector.start_heartbeat(self.hb)
+
+                logger.debug(f"Connector status is {connector.status}")
+
+                # we trigger a sync
+                if connector.status in (Status.CREATED, Status.NEEDS_CONFIGURATION):
+                    # we can't sync in that state
+                    logger.info(f"Can't sync with status `{e2str(connector.status)}`")
+                    return
+
+                if not connector.sync_now:
+                    next_sync = connector.next_sync()
+                    if next_sync == SYNC_DISABLED or next_sync - self.idling > 0:
+                        if next_sync == SYNC_DISABLED:
+                            logger.debug(
+                                f"Scheduling is disabled for {connector.service_type}"
+                            )
+                        else:
+                            logger.debug(
+                                f"Next sync for {connector.service_type} due in {int(next_sync)} seconds"
+                            )
+                        # if we don't sync, we still want to make sure we tell kibana we are connected
+                        # if the status is different from connected
+                        if connector.status != Status.CONNECTED:
+                            connector.status = Status.CONNECTED
+                            await connector.sync_doc()
+                        return
+                else:
+                    logger.info("Sync forced")
+
+                await self._create_job(connector)
+
+                await asyncio.sleep(0)
+            finally:
+                await connector.close()
+
+    async def _create_job(self, connector):
+        trigger_method = (
+            JobTriggerMethod.ON_DEMAND
+            if connector.sync_now
+            else JobTriggerMethod.SCHEDULED
+        )
+
+        job_id = await self.sync_job_index.create(connector, trigger_method)
+
+        connector.sync_now = connector.doc_source["sync_now"] = False
+
+        await connector.sync_doc()
+
+        logger.info(
+            f"Created a job for connector {connector.service_type}, Job id: {job_id}"
+        )
+
+    async def run(self):
+        if "PERF8" in os.environ:
+            import perf8
+
+            async with perf8.measure():
+                return await self._run()
+        else:
+            return await self._run()
```

Comment:
I'm not sure PERF8 is needed here - but if so, will the measurements from this service go to a different directory than the job executor? It's nice to have these metrics separated cc @tarekziade

Reply:
I had the similar question when I was drafting the PR: should `trace_mem` only be applied to job execution? But I see `trace_mem` was removed by another PR, so the question now is: should `PERF8` be applied at the application level, service level (i.e. job scheduler service and job executor service) or job execution level?

Reply:
I think we can put this back into the `cli.py` top-level function because we measure these things in the context of a whole benchmark. We do want to capture interactions and problems between services. For instance the event loop lag.

Reply:
I have done it in this PR: https://github.com/elastic/connectors-python/pull/337